File: benchmarks/src/python/method_dispatch.py
import time

class Shape:
    def score(self):
        return 0

class A(Shape):
    def score(self):
        return self.v * 2

class B(Shape):
    def score(self):
        return self.v * 3 + 1

class C(Shape):
    def score(self):
        return self.v * 5 - 2

shapes = []
for i in range(3000):
    r = i % 3
    if r == 0:
        obj = A()
    elif r == 1:
        obj = B()
    else:
        obj = C()
    obj.v = i % 97
    shapes.append(obj)

P = 300
acc = 0
start = time.time() * 1000
for _ in range(P):
    for s in shapes:
        acc = (acc + s.score()) % 1000003
end = time.time() * 1000

print(f"Result: {acc}")
print(f"Time: {int(end - start)}ms")

File: benchmarks/src/php/method_dispatch.php
<?php
class Shape { function score() { return 0; } }
class A extends Shape { public $v = 0; function score() { return $this->v * 2; } }
class B extends Shape { public $v = 0; function score() { return $this->v * 3 + 1; } }
class C extends Shape { public $v = 0; function score() { return $this->v * 5 - 2; } }

$shapes = [];
for ($i = 0; $i < 3000; $i++) {
    $r = $i % 3;
    if ($r == 0) $obj = new A();
    elseif ($r == 1) $obj = new B();
    else $obj = new C();
    $obj->v = $i % 97;
    $shapes[] = $obj;
}

$P = 300;
$acc = 0;
$start = hrtime(true);
for ($round = 0; $round < $P; $round++) {
    foreach ($shapes as $s) {
        $acc = ($acc + $s->score()) % 1000003;
    }
}
$end = hrtime(true);
$ms = ($end - $start) / 1e6;
echo "Result: " . $acc . "\n";
echo "Time: " . round($ms) . "ms\n";
